feat: adding data processing script

isond opened the pull request:
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
What is the larger goal of this change?
To add the data processing script that will organize the NHTS data.

### What the code is doing
How is the purpose executed?
This code filters through the NHTS data based on the inputted census division and reorganizes the information. It also calculates additional information based on the given data.

### Testing
How did you test this change (unit/functional testing, manual testing, etc.)? 
This was tested through manual testing using data from "trippub.csv".

### Where to look
* It's helpful to clarify where your new code lives if you moved files around or there could be confusion/
The new code is located at "PreREISE/prereise/gather/demanddata/transportation_electrification/data_process.py"

* What files are most important?
data_process.py which is currently the only file

### Usage Example/Visuals
How the code can be used and/or images of any graphs, tables or other visuals (not always applicable). 

### Time estimate
How long will it take for reviewers and observers to understand this code change?
It should take a few days for reviewers to understand.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+
+    #dataframe for use built here
+    #columns taken directly from original csv:
+    sorted_data = pd.DataFrame({"Household": nhts["HOUSEID"], 
+                   "Vehicle ID": nhts["VEHID"], 
+                   "Person ID": nhts["PERSONID"],
+                   "Scaling Factor Applied": nhts["WTTRDFIN"],
+                   "Trip Number": nhts["TDTRPNUM"],
+                   "Date": nhts["TDAYDATE"], 
+                   "Day of Week": nhts["TRAVDAY"],              
+                   "If Weekend": nhts["TDWKND"],                
+                   "Trip start time (HHMM)": nhts["STRTTIME"],
+                   "Trip end time (HHMM)": nhts["ENDTIME"],
+                   "Travel Minutes": nhts["TRVLCMIN"],
+                   "Dwell time": nhts["DWELTIME"],
+                   "Miles traveled": nhts["VMT_MILE"],          
+                   "Vehicle miles travelled": nhts["TRPMILES"], 
+                   "why from": nhts["WHYFROM"],
+                   "why to": nhts["WHYTO"],
+                   "Vehicle type": nhts["TRPTRANS"],            
+                   "Household vehicle count": nhts["HHVEHCNT"],
+                   "Household size": nhts["HHSIZE"]
+                   })
```
Since `nhts` is already a DataFrame, I think we can simplify this using the `rename` method. So instead of:
```python
sorted_data = pd.DataFrame({"oldname1": nhts["newname1"], "oldname2": nhts["newname2"], ...})
```
we could do something like:
```python
column_translations = {"oldname1": "newname1", "oldname2": "newname2", ...}
sorted_data = nhts[column_translations.keys()].rename(column_translations, axis="columns")
```

rouille replied:
Or we can name overwrite column names using `header=0` and the `names` parameter in `read_csv`.

danielolsen replied:
I think if we use the `names` parameter, we'll need to pass a list of names that cover all of the columns, even though we're only interested in a subset of them.

rouille replied:
And using `usecols` together with `names`? The nice thing with `usecols` is that it can take à `range` or integers.

isond replied:
I decided to go with Daniel's first suggestion since not all the columns in the original data set have an alternative name they could map to if we use `names`. But if I overlooked how to use it, let me know and I'll try to implement it!

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+def data_filtering(census_division, filepath: str = "trippub.csv"):
+    """Filter raw NHTS data to be used in mileage.py
+    
+    :param str filepath: filepath for the "trippub.csv" file in downloaded NHTS data
```
Missing docstring parameter for the `census_division` input.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+    #columns that require calculations: 21-28
+    sorted_data["Start time (hour decimal)"] = [row//100 + row % 100 / 60 for row in sorted_data["Trip start time (HHMM)"]]
+    sorted_data["End time (hour decimal)"] = [row//100 + row % 100 / 60 for row in sorted_data["Trip end time (HHMM)"]]
+    sorted_data['Dwell time (hour decimal)'] = ''
```
If all we're doing here is expanding the set of columns without filling meaningful data, then instead of:
```python
df["a"] = '',
df["b"] = '',
...
```
we can add multiple columns at once without them being set to string/object type with something like:
```python
df = df.reindex([df.columns + ["a", "b", ...], axis="columns")
```

danielolsen commented:
As a note for BE folks: the target for this PR is currently the `develop` branch, so that we can enable the CI linting, but we will want to change the target to the `transportation_electrification` branch before merging.

rouille commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+    :return: (*pd.DataFrame*) -- filtered and sorted trip data
+    """
+
+    nhts_raw = pd.read_csv(filepath)
```
@BainanXia and @danielolsen, I believe we want this file on our blob storage, read from there and refer to it in the **ATTRIBUTION** file. I believe, we could have a year parameter that would get the file for a given year and raise an error if it is not available on the blob storage.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+
+        i += 1
+
+    sorted_data.to_excel("output.xlsx")
```
We probably want to make this step optional, since I don't think we use Excel files downstream.

isond replied:
This was actually more for me when testing things. I can go ahead and remove it

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+
+    return sorted_data
+
+# data_filtering(1)
```
We'll want to remove this comment.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+
+#Download CSV of NHTS 2017 from https://nhts.ornl.gov/
+
+def data_filtering(census_division, filepath: str = "trippub.csv"):
```
If we refactor this function to take a dataframe instead of a filepath, it will be easier to write automated tests which operate only on objects in-memory.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,132 @@
+    total_trips = 0
+    total_miles = 0
+
+    while i < l:
```
As @limingzhou2004 points out, we can also combine the calculation of group totals with the assignment back to individual rows:
```python
total_trips = grouping.size()
df["total_trips"] = df["class"].map(total_trips)
total_miles = grouping["miles"].sum()
df["total_miles"] = df["class"].map(total_miles)
```
can become:
```python
df["total_trips"] = grouping["miles"].transform(len)
df["total_miles"] = grouping["miles"].transform(sum)
```
This will avoid us having to define a series of `keys` values.

danielolsen commented:
After the refactor, the code runs much faster: less than 4 minutes now to go through all nine regions on my laptop, rather than the hour and a half before. Bravo!

Comparing the results from before the refactor to the results afterward, I see a few differences:
- Some column renaming for spelling (`travelled` -> `traveled`)
- Different lengths of the dataframes for each region

| Region | Length before refactor | Length after refactor |
| ------- | ------------------------- | ----------------------- |
| 1 | 13,323 | 9,030 |
| 2 | 123,616 | 82,355 |
| 3 | 104,046 | 73,197 |
| 4 | 35,424 | 24,964 |
| 5 | 192,380 | 136,168 |
| 6 | 8,424 | 5,999 |
| 7 | 183,089 | 130,259 |
| 8 | 34,968 | 23,520 |
| 9 | 189,215 | 122,217 |

If I look at the indices from the newly-returned dataframes, and look at the corresponding values in the previously-returned dataframes, I see that most columns are the same, some columns have values that are different but nearly-identical: `Scaling Factor Applied`, `Miles traveled`, `Vehicle miles traveled`, `Start time (hour decimal)`, `End time (hour decimal)`, `Travel time (hour decimal)`, `Vehicle speed (mi/hour)` (probably differences due to floating-point rounding across the two different methods of summing), and some columns have values that are noticeably different: `Dwell time (hour decimal)`, `sample vehicle number`, `total vehicle trips`, and `total vehicle miles traveled`.

@isond can you confirm that the behavior change is intentional, and that the new behavior is the one we want?

One way to validate the new behavior could be to refactor the `data_filtering` function to take the raw NHTS dataframe as an explicit input, rather than to load it from disk. So instead of the user getting data as:
```python
df1 = data_filtering(census_division=1)
```
we could:
```python
raw_nhts = pd.read_csv(FILENAME)
df1 = data_filtering(raw_nhts, census_division=1)
```
The value of an approach like this is that we can design a test that passes a very small amount of data to `data_filtering`, and compares the output to a known expectation (calculated manually). Another benefit is that we can run the `data_filtering` function from anywhere, and we're less tied to specific filenames; we won't need the current working directory of the script to contain a file named `trippub.csv`.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,133 @@
+    if not (1 <= census_division <= 9):
+        raise ValueError("census_region must be between 1 and 9 (inclusive).")
+
+    nhts_census = pd.read_csv(filepath)
+
+    df = pd.DataFrame(nhts_census)
```
`pd.read_csv` already returns a `DataFrame`, so we don't need the `df = pd.DataFrame(nhts_census)` line, we could just do:
```python
nhts_census = pd.read_csv(filepath)
return nhts_census
```
or even:
```python
return pd.read_csv(filepath)
```

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,133 @@
+    :param int census_division: any of the 9 census regions defined by the US census bureau
+    :param str filepath: filepath for the "trippub.csv" file in downloaded NHTS data
+    :raises ValueError: if the census division is not between 1 and 9, inclusive.
+    :return: (*pd.DataFrame*) -- the data loaded from trippub.csv.
```
In docstrings and type annotations, we still want to use `pandas.DataFrame` rather than `pd.DataFrame`.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,133 @@
+
+
+def data_filtering(census_division):
+    """Filter raw NHTS data to be used in mileage.py
+
+    :param int census_division: any of the 9 census regions defined by the US census bureau
+    :param str filepath: filepath for the "trippub.csv" file in downloaded NHTS data
+    :return: (*pd.DataFrame*) -- filtered and sorted trip data
+    """
```
Since this is the main user-facing function, we want to make sure that the docstring here has the right input parameters listed.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,133 @@
+
+    nhts_raw = load_data(census_division)
+
+    # filter to be only vehicle trips (TRPTRANS values 1-6)
+    nhts = nhts_raw[nhts_raw.TRPTRANS.isin(range(1, 7))]
+
+    # filter out repeated trips (VMT = -1)
+    nhts = nhts[nhts.VMT_MILE != -1]
+
+    # get correct census division
+    nhts = nhts[nhts.CENSUS_D == census_division]
```
Since each of these comparisons produces a binary array, they could be chained together as:
```python
nhts = nhts_raw.loc[
    nhts_raw.TRPTRANS.isin(range(1, 7))
    & (nhts.VMT_MILE != -1)
    & (nhts.CENSUS_D == census_division)
]
```
Both are equivalent though, so it's up to you which version you prefer.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,133 @@
+        "HHSIZE": "Household size",
+    }
+
+    sorted_data = nhts[column_translations.keys()].rename(
+        column_translations, axis="columns"
+    )
```
I'm not sure I understand this naming: are we actually sorting the data?

isond replied:
I guess it's more like filtering the data; hence the function name haha

isond commented:
> After the refactor, the code runs much faster: less than 4 minutes now to go through all nine regions on my laptop, rather than the hour and a half before. Bravo!
> 
> Comparing the results from before the refactor to the results afterward, I see a few differences:
> 
> * Some column renaming for spelling (`travelled` -> `traveled`)
> * Different lengths of the dataframes for each region
> 
> Region	Length before refactor	Length after refactor
> 1	13,323	9,030
> 2	123,616	82,355
> 3	104,046	73,197
> 4	35,424	24,964
> 5	192,380	136,168
> 6	8,424	5,999
> 7	183,089	130,259
> 8	34,968	23,520
> 9	189,215	122,217
> If I look at the indices from the newly-returned dataframes, and look at the corresponding values in the previously-returned dataframes, I see that most columns are the same, some columns have values that are different but nearly-identical: `Scaling Factor Applied`, `Miles traveled`, `Vehicle miles traveled`, `Start time (hour decimal)`, `End time (hour decimal)`, `Travel time (hour decimal)`, `Vehicle speed (mi/hour)` (probably differences due to floating-point rounding across the two different methods of summing), and some columns have values that are noticeably different: `Dwell time (hour decimal)`, `sample vehicle number`, `total vehicle trips`, and `total vehicle miles traveled`.
> 
> @isond can you confirm that the behavior change is intentional, and that the new behavior is the one we want?
> 
> One way to validate the new behavior could be to refactor the `data_filtering` function to take the raw NHTS dataframe as an explicit input, rather than to load it from disk. So instead of the user getting data as:
> 
> ```python
> df1 = data_filtering(census_division=1)
> ```
> 
> we could:
> 
> ```python
> raw_nhts = pd.read_csv(FILENAME)
> df1 = data_filtering(raw_nhts, census_division=1)
> ```
> 
> The value of an approach like this is that we can design a test that passes a very small amount of data to `data_filtering`, and compares the output to a known expectation (calculated manually). Another benefit is that we can run the `data_filtering` function from anywhere, and we're less tied to specific filenames; we won't need the current working directory of the script to contain a file named `trippub.csv`.

Regarding the differences in lengths of the dataframes, I noticed that the condition to filter out the repeated trips was using the incorrect column prior to the refactor. 

`nhts = nhts[nhts.TRPMILES != -1]` should be `nhts = nhts[nhts.VMT_MILE != -1]`

Basically in the original code, the two columns were switched around, and so I've updated the code to reflect that change. 
I manually filtered out the columns in the `trippub.csv` because I wanted to double-check the length of the results from the dataframes after the refactor, and they all match up for each region. So the lengths that you are seeing now should be good :)

The significant change in values for `Dwell time (hour decimal)`, `sample vehicle number`, `total vehicle trips`, and `total vehicle miles traveled` is due to the fact that they used `vehicle miles traveled` (originally using values from `TRPMILES` but is now using `VMT_MILE`) which was affected by the change I previously described.

As for the spelling change, I updated the "traveled" variable to be consistent with the variables listed in Kate's data conversion document.

danielolsen commented on prereise/gather/demanddata/transportation_electrification/data_process.py:
```diff
@@ -0,0 +1,113 @@
+def data_filtering(raw_nhts, census_division):
+    """Filter raw NHTS data to be used in mileage.py
+
+    :param: (*pandas.DataFrame*) raw_nhts: raw NHTS dataframe
```
`:param: (*pandas.DataFrame*) raw_nhts:` -> `:param pandas.DataFrame raw_nhts:`. We only add the parentheses for the return variable.

danielolsen commented:
All checks pass, and  the code runs successfully and produces dataframes of the appropriate length, in 2 minutes:
```python
import pandas as pd
from prereise.gather.demanddata.transportation_electrification.data_process import data_filtering
raw_nhts = pd.read_csv("trippub.csv")
data = {i: data_filtering(raw_nhts, i) for i in range(1, 10)}
```

There's one small tweak that needs to be made to a docstring, and the branch needs to be rebased onto the latest `transportation_electrification` commit, and then we're ready to merge!

danielolsen reviewed (approved):
Looks good to me!
